src/pykeen/sampling/negative_sampler.py

```diff
@@ -44,7 +51,50 @@ def num_entities(self) -> int:  # noqa: D401
         """The number of entities to sample from."""
         return self.triples_factory.num_entities
 
+    @property
+    def num_relations(self) -> int:  # noqa: D401
+        """The number of relations to sample from."""
+        return self.triples_factory.num_relations
+
     @abstractmethod
-    def sample(self, positive_batch: torch.LongTensor) -> torch.LongTensor:
+    def sample(self, positive_batch: torch.LongTensor) -> Tuple[torch.LongTensor, Optional[torch.Tensor]]:
         """Generate negative samples from the positive batch."""
         raise NotImplementedError
+
+    def _filter_negative_triples(self, negative_batch: torch.LongTensor) -> torch.Tensor:
+        """Filter all proposed negative samples that are positive in the training dataset.
+
+        Normally there is a low probability that proposed negative samples are positive in the training datasets and
+        thus act as false negatives. This is expected to act as a kind of regularization, since it adds noise signal to
+        the training data. However, the degree of regularization is hard to control since the added noise signal depends
+        on the ratio of true triples for a given entity relation or entity entity pair. Therefore, the effects are hard
+        to control and a researcher might want to exclude the possibility of having false negatives in the proposed
+        negative triples.
+        """
+        # Make sure the mapped triples are on the right device
+        if not self._filter_init:
+            # Copy the mapped triples to the device for efficient filtering
+            self.mapped_triples = self.triples_factory.mapped_triples.to(negative_batch.device)
+            self._filter_init = True
+        try:
+            # Check which heads of the mapped triples are also in the negative triples
+            head_filter = (self.mapped_triples[:, 0:1].view(1, -1) == negative_batch[:, 0:1]).max(axis=0)[0]
+            # Reduce the search space by only using possible matches that at least contain the head we look for
+            sub_mapped_triples = self.mapped_triples[head_filter]
+            # Check in this subspace which relations of the mapped triples are also in the negative triples
+            relation_filter = (sub_mapped_triples[:, 1:2].view(1, -1) == negative_batch[:, 1:2]).max(axis=0)[0]
```

Comment:
the `[0]` index at the end is a bit opaque. the result of max(axis=0) a 2-tuple, right? I'd prefer it to be shown as assignment with tuple unpacking like
```suggestion
            relation_filter, _ = (sub_mapped_triples[:, 1:2].view(1, -1) == negative_batch[:, 1:2]).max(axis=0)
```

I know this might cause another assignment, but it will promote understandability for the code later especially for someone like me who can't remember what the return type from max() is

Reply:
There are a couple other places with this kind of assignment. If you agree this change for readability isn't going to cause a performance hit, it would be good to make the same change in other places using similar indexing

Reply:
I agree on code readability, but the assignment of the variable will increase the memory pressure.

Reply:
Therefore, I would like to avoid doing that.

Reply:
Fair enough - best to leave it as is then